src/OrchardCore/OrchardCore.Sms.Core/ServiceCollectionExtensions.cs

```diff
@@ -10,54 +11,49 @@ public static class ServiceCollectionExtensions
 {
     public static IServiceCollection AddSmsServices(this IServiceCollection services)
     {
-        services.AddSingleton<DefaultSmsProvider>();
-        services.AddTransient<IPostConfigureOptions<SmsSettings>, SmsSettingsConfigurator>();
-        services.AddSingleton(sp =>
+        services.AddTransient<IPostConfigureOptions<SmsSettings>, SmsSettingsConfiguration>();
+        services.AddSingleton(serviceProvider =>
         {
-            var settings = sp.GetRequiredService<IOptions<SmsSettings>>().Value;
+            var settings = serviceProvider.GetRequiredService<IOptions<SmsSettings>>().Value;
 
             if (!String.IsNullOrEmpty(settings.DefaultProviderName))
             {
-                var smsProviderOptions = sp.GetRequiredService<IOptions<SmsProviderOptions>>().Value;
+                var smsProviderOptions = serviceProvider.GetRequiredService<IOptions<SmsProviderOptions>>().Value;
 
                 if (smsProviderOptions.Providers.TryGetValue(settings.DefaultProviderName, out var providerGetter))
                 {
-                    return providerGetter(sp);
+                    return providerGetter(serviceProvider);
```

Comment:
You don't need to enlist getters in the options, only types and then use `CreateInstance()` here.

Reply:
I thought about that. The problem is that we expose the directory in the SmsProviderOptions. someone can register an invalid types in the directory not just a Type that implements the `ISmsProvider`

another thing we can do is 
```
public class SmsProviderOptions
{
    public Dictionary<string, ISmsProvider> Providers { get; } = new();
}
```

Then instead of the getter we can do something like `serviceProvider.CreateInstance(provider.GetType())`

Reply:
No problem to enlist types and `AddSmsProvider()` has a constraint on `ISmsProvider`.